torch/_functorch/autograd_function.py

```diff
@@ -720,33 +720,68 @@ def __call__(self, fwd, bwd, *fwd_args, **fwd_kwargs):
         # since we don't need to calculate the gradients of them.
         new_fwd_args = fwd_args[:length_of_tensor_args]
 
-        class ApplyTemplate(torch.autograd.Function):
-            @staticmethod
-            def forward(ctx, *args):
-                nonlocal saved_values
-                output, saved_values = fwd(None, *fwd_args)
+        # If users call ctx.mark_non_differentiable() in the original fwd function.
+        def _ctx_mark_non_differentiable(ctx, output, non_differentiable_idx):
+            if len(non_differentiable_idx) > 0:
+                non_differentiable_output = []
+                for i, x in enumerate(output):
+                    if i in non_differentiable_idx:
+                        non_differentiable_output.append(x)
+                ctx.mark_non_differentiable(*non_differentiable_output)
 
-                # If users call ctx.mark_non_differentiable() in the original fwd function.
-                if len(non_differentiable_idx) > 0:
-                    non_differentiable_output = []
-                    for i, x in enumerate(output):
-                        if i in non_differentiable_idx:
-                            non_differentiable_output.append(x)
-                    ctx.mark_non_differentiable(*non_differentiable_output)
+        if is_setup_ctx_defined:
 
-                return output
+            class ApplyTemplate(torch.autograd.Function):
+                nonlocal original_generate_vmap_rule
+                generate_vmap_rule = original_generate_vmap_rule
 
-            @staticmethod
-            def backward(ctx, *grad):
-                return bwd(None, *grad, *saved_values)
+                @staticmethod
+                def forward(*args):
+                    nonlocal saved_values
+                    output, saved_values = fwd(None, *fwd_args)
+                    return output
 
-        return ApplyTemplate.apply(*new_fwd_args)
+                @staticmethod
+                def setup_context(ctx, inputs, output):
+                    nonlocal saved_values
+                    ctx.save_for_backward(*saved_values)
+
+                    _ctx_mark_non_differentiable(ctx, output, non_differentiable_idx)
+
+                @staticmethod
+                def backward(ctx, *grad):
+                    saved_values = ctx.saved_tensors
+                    return bwd(None, *grad, *saved_values)
+
+            return ApplyTemplate.apply(*new_fwd_args)
+        else:
+
+            class ApplyTemplate(torch.autograd.Function):  # type: ignore[no-redef]
+                nonlocal original_generate_vmap_rule
+                generate_vmap_rule = original_generate_vmap_rule
+
+                @staticmethod
+                def forward(ctx, *args):
+                    nonlocal saved_values
+                    output, saved_values = fwd(None, *fwd_args)
+
+                    _ctx_mark_non_differentiable(ctx, output, non_differentiable_idx)
```

Comment:
This shouldn't be necessary, right? generate_vmap_rule should be false in this case (we should assert otherwise). IIRC generate_vmap_rule requires a split forward